[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132931

We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"

We need to make changes to the test to make it aligned with the change.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @gchanan

fduwjj committed:
[c10d] Expose backend Options to reflect the correct code structure

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Expose backend Options to reflect the correct code structure"


What I noticed is that we still assume PGNCCL is using option from ProcessGroup but we actually should make it use Option from backend class.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1533,7 +1533,7 @@ def init_process_group(
             store,
             group_name,
-            pg_options=pg_options,
+            backend_options=pg_options,
```
Since a process group can have multiple backends, I think we should discuss how users will provide backend options for the multiple backends that are created. I think we may need a dict mapping device->option for the backends that are created?

Another question to resolve is how will this work for current users that are passing in the existing process group options. What is the transition plan for that and how can we provide sensible defaults?

fduwjj replied:
Well this PR is just for renaming. How should we continue this discussion down the road? Also I remember @shuqiangzhang is asking if we really have use cases when multiple backends that are created.

fduwjj replied:
I think let's discuss an ideal behavior first and then we can see what transition plan we can have here.

fduwjj committed:
Update on "[c10d] Expose backend Options to reflect the correct code structure"


Although it looks like documents for Options mentioned in https://github.com/pytorch/pytorch/issues/130958 have been all addressed. What I noticed is that we still assume PGNCCL is using option from ProcessGroup but we actually should make it use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

wz337 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4601,6 +4609,7 @@ def _new_group_with_tag(
     if not backend:
         backend = default_backend
     backend = Backend(backend)
+    print(backend)
```
ditto

wz337 commented on torch/distributed/device_mesh.py:
```diff
@@ -34,6 +34,7 @@ def _init_device_mesh_stub():
 
 
 else:
+    from torch._C._distributed_c10d import Backend as C10dBackend
```
Thanks for handling this.

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"

We need to make changes to the test to make it aligned with the change.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1704,11 +1705,15 @@ def _new_process_group_helper(
+        group_size,
     )
+    assert backend in Backend.backend_type_map, f"Unknown backend type {backend}"
+    pg._set_default_backend(Backend.backend_type_map[backend])
```
if backend isn't provided (e.g. `backend=None`), then the backend type will be undefined right? Will this break the APIs that depend on the looking at the backend type, like getSequenceNumber or the hooks that we talked about.

I'm not sure if we have tests that cover these cases, but could be worthwhile to check.

fduwjj replied:
The backend is not optional for this function. Do we see the use case when `backend=None`? This logic will break when the backend is something like "cpu:gloo,gpu:nccl".

H-Huang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2556,6 +2541,27 @@ options :class:`~torch.distributed.ProcessGroupNCCL.Options`).
               &::c10d::Backend::endCoalescing,
               py::call_guard<py::gil_scoped_release>());
 
+  // base Backend::Options binding
```
Maybe add a TODO for thhe consolidation of this with `DistributedBackendOptions` that we discussed before?

H-Huang reviewed (approved):
Overall LGTM on the process group changes. I would probably test that optional backend in `init_proces_group` still work with `TORCH_DISTRIBUTED_DEBUG=DETAIL` since I'm not sure how the backendType_ will be recognized.

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"

We need to make changes to the test to make it aligned with the change.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

fduwjj commented:
Thanks for your quick review, I remember there are unit tests for the case when `TORCH_DISTRIBUTED_DEBUG=DETAIL`?

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"

We need to make changes to the test to make it aligned with the change.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Remove Option for ProcessGroup and Expose backend Options to reflect the correct code structure"


We introduced the dispatchable backend for a ProcessGroup and collective in https://github.com/pytorch/pytorch/issues/86225. This PR is a follow-up cleanup to clean up the option of a ProcessGroup and ask users to either set timeout or backend later on or directly create backend after creating a PG.

Also PGNCCL is using option class from ProcessGroup but we actually should use Option from backend class. So this PR is to make the type or name to be aligned with what we are doing in cpp side. I don't change the signature for the public API, so they still use args named "pg_options"

We need to make changes to the test to make it aligned with the change.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o ezyang gchanan

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge -f "The failed one is trunk test, not related"

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "This PR is breaking builds internally due to the removal of ProcessGroup::Options"

I'd recommend ghimporting this PR to make sure it passes internally before re-merging.  See D62008954 for more details

fduwjj commented:
reland in https://github.com/pytorch/pytorch/pull/135653
